### DISCUSSION TITLE
(Optionally) defer user creation until verified

### DISCUSSION THREAD

**Discussion by Luthien-in-edhil** [keycloak/keycloak#14877] [Ideas]
Related to issue [11800 (Automatically disable dormant users)](https://github.com/keycloak/keycloak/issues/11800) and discussion [#9348 (Disable inactive users after n number of days) ](https://github.com/keycloak/keycloak/discussions/9348) I'd like to suggest having the option to defer creating a user account until it has been verified.

I understand that there is an overhead to keeping track of user activity and making sure accounts are not pruned without a good reason, but wouldn't it be much less complex to defer creating the account until the verification step has been completed within a configurable time slot? 
Then, if the time slot passes without validating, the account never gets created in the first place.

This would stop the cluttering of unused user accounts (10% of all in our case) and block potential abuse by users blocking a given email address or username from being used to register (see also [this similar discussion](https://gitlab.com/gitlab-org/gitlab/-/issues/352514) on Gitlab).

Alternatively, it would also be helpful if it was possible to search for non-validated users in the Admin UI and batch delete those. 
It's of course possible to query the Admin Rest API for unverified users and delete the resulting users by their ID, but that's a cumbersome workflow, requiring setting up  - for instance - something in Insomnia or Postman to the effect of:

1. obtaining the right token 
2. using that to query unverified users 
3. extracting the UserID from the response, and 
4. issuing the delete command 

Adding one button in the UI could accomplish the same thing, and make a lot of users love Keycloak even better.

**Comment by upfluentpatrickpilch** [comment]
I am trying to leverage Keycloak for our user registration and I'm running into this difficulty as well. 

The out of the box registration flow is difficult to alter for a "multi-step" registration flow, e.g. collecting contact information like phone/email and verifying it _**before**_ creation of a user. You end up having to re-implement a lot of the logic contained in the registration `FormAction`s due to the assumption that the "Registration User Creation" form action always comes first with a single form containing almost all the user info. 

Without a more flexible in-progress registration flow approach, I share the same concern of a denial of service attack using up emails and creating a maintenance burden.
